File: scripts/build_data.py
import json
import re
from datetime import datetime, timezone
from pathlib import Path

APP_DIR = Path(__file__).resolve().parents[1]
ROOT_DIR = APP_DIR.parents[1]
LANGUAGES = ["English", "Japanese", "Chinese"]


def main():
    data = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "daily": {},
        "issues": {},
        "vocabulary": {},
    }
    for language in LANGUAGES:
        data["daily"][language] = read_lesson_folder("01_Daily Sentences", language, parse_daily)
        data["issues"][language] = read_lesson_folder("02_World Issues", language, parse_issue)

    out_dir = APP_DIR / "data"
    out_dir.mkdir(exist_ok=True)
    (out_dir / "lessons.json").write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


def read_lesson_folder(section, language, parser):
    folder = ROOT_DIR / section / language
    return [parser(path.read_text(encoding="utf-8"), path.name, language) for path in sorted(folder.glob("*.md"))]


def parse_daily(raw, file_name, language):
    meta = read_frontmatter(raw)
    body = strip_frontmatter(raw)
    sentence_section = between(body, "## Sentences", "## 오늘 바로 써볼 문장")
    if not sentence_section:
        sentence_section = re.sub(r"^# .+\n*", "", body).split("## 오늘 바로 써볼 문장")[0].strip()

    sentences = []
    for block in [part.strip() for part in re.split(r"\n(?=\d+\.\s)", sentence_section) if part.strip()]:
        lines = [line.strip() for line in block.splitlines() if line.strip()]
        text_lines = []
        for line in lines:
            if line.startswith(("한국어:", "고난이도", "- ")):
                break
            text_lines.append(re.sub(r"^\d+\.\s*", "", line))
        reading, original = split_reading_and_original(text_lines, language)
        korean = next((line.replace("한국어:", "", 1).strip() for line in lines if line.startswith("한국어:")), "")
        expressions = extract_bullet_expressions(block)
        sentences.append({"reading": reading, "original": original, "korean": korean, "expressions": expressions})

    practice_match = re.search(r"-\s*(.+)", between(body, "## 오늘 바로 써볼 문장", ""))
    date = meta.get("date") or date_from_file(file_name)
    return {
        "date": date,
        "language": language,
        "title": f"{language} Daily Sentences - {date}",
        "practice": practice_match.group(1) if practice_match else (sentences[0]["original"] if sentences else ""),
        "sentences": sentences,
        "expressions": [expr for sentence in sentences for expr in sentence["expressions"]],
        "sourceFile": file_name,
    }


def parse_issue(raw, file_name, language):
    meta = read_frontmatter(raw)
    body = strip_frontmatter(raw)
    issue_section = between(body, "## Issue", "## 내 의견 한 줄")
    korean_match = re.search(r"한국어:\s*(.+)", issue_section)
    korean = korean_match.group(1) if korean_match else ""
    expressions = extract_bullet_expressions(issue_section)
    opinion_section = between(body, "## 내 의견 한 줄", "")
    practice_match = re.search(r"-\s*(?:English|Japanese|Chinese):\s*(.+)", opinion_section) or re.search(r"-\s*한국어:\s*(.+)", opinion_section)
    lines = [line.strip() for line in issue_section.splitlines() if line.strip()]
    text_lines = [
        line for line in lines
        if not line.startswith(("한국어:", "고난이도", "-", "Sources:"))
    ]
    reading, original = split_reading_and_original(text_lines, language)
    date = meta.get("date") or date_from_file(file_name)
    return {
        "date": date,
        "language": language,
        "title": f"{language} World Issue - {date}",
        "topic": meta.get("topic", ""),
        "practice": practice_match.group(1) if practice_match else korean,
        "issueCards": [{
            "title": meta.get("topic", "") or "World Issue",
            "reading": reading,
            "original": original or korean,
            "korean": korean,
            "expressions": expressions,
        }],
        "expressions": expressions,
        "sources": [line for line in issue_section.splitlines() if line.strip().startswith("- [")],
        "sourceFile": file_name,
    }


def read_frontmatter(raw):
    match = re.match(r"^---\n([\s\S]*?)\n---", raw)
    if not match:
        return {}
    meta = {}
    for line in match.group(1).splitlines():
        if ":" in line:
            key, value = line.split(":", 1)
            meta[key.strip()] = value.strip()
    return meta


def strip_frontmatter(raw):
    return re.sub(r"^---\n[\s\S]*?\n---\n*", "", raw)


def between(text, start_heading, end_heading):
    start = text.find(start_heading)
    if start == -1:
        return ""
    body_start = start + len(start_heading)
    end = text.find(end_heading, body_start) if end_heading else -1
    return text[body_start:(len(text) if end == -1 else end)].strip()


def extract_bullet_expressions(block):
    expressions = []
    current = None
    for raw_line in block.splitlines():
        line = raw_line.strip()
        if line.startswith("- Hiragana:") or line.startswith("- Pinyin:"):
            if current:
                expressions.append(normalize_expression(current))
            label, _, value = line[2:].partition(":")
            current = {"readingLabel": label.strip(), "reading": value.strip()}
            continue

        if current and line.startswith(("표현:", "한국어:", "활용:")):
            label, _, value = line.partition(":")
            current[label.strip()] = value.strip()
            continue

        if not line.startswith("- "):
            continue
        value = line[2:]
        if value.startswith("["):
            continue
        if current:
            expressions.append(normalize_expression(current))
            current = None
        term, _, meaning = value.partition(":")
        term = clean_label(term)
        if term:
            expressions.append({"term": term, "meaning": clean_label(meaning)})
    if current:
        expressions.append(normalize_expression(current))
    return expressions


def normalize_expression(item):
    term = item.get("표현") or item.get("term") or ""
    meaning = item.get("한국어") or item.get("meaning") or ""
    note = item.get("활용") or ""
    return {
        "term": clean_label(term),
        "reading": clean_label(item.get("reading", "")),
        "meaning": clean_label(meaning),
        "note": clean_label(note),
    }


def clean_label(value):
    return re.sub(r"^(Hiragana|Pinyin|표현|한국어|활용):\s*", "", value or "").strip()


def split_reading_and_original(text_lines, language):
    text_lines = [line for line in text_lines if line]
    if language in {"Japanese", "Chinese"} and len(text_lines) >= 2:
        return text_lines[0], text_lines[1]
    return "", text_lines[0] if text_lines else ""


def date_from_file(file_name):
    match = re.search(r"\d{4}-\d{2}-\d{2}", file_name)
    return match.group(0) if match else ""


if __name__ == "__main__":
    main()

File: scripts/build-data.mjs
import { mkdir, readdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const appDir = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const rootDir = path.resolve(appDir, "../..");
const languages = ["English", "Japanese", "Chinese"];

const data = {
  generatedAt: new Date().toISOString(),
  daily: {},
  issues: {},
  vocabulary: {}
};

for (const language of languages) {
  data.daily[language] = await readLessonFolder("01_Daily Sentences", language, parseDaily);
  data.issues[language] = await readLessonFolder("02_World Issues", language, parseIssue);
}

await mkdir(path.join(appDir, "data"), { recursive: true });
await writeFile(path.join(appDir, "data", "lessons.json"), JSON.stringify(data, null, 2));

async function readLessonFolder(section, language, parser) {
  const dir = path.join(rootDir, section, language);
  const files = (await readdir(dir)).filter((file) => file.endsWith(".md")).sort();
  const lessons = [];
  for (const file of files) {
    const raw = await readFile(path.join(dir, file), "utf8");
    lessons.push(parser(raw, file, language));
  }
  return lessons;
}

function parseDaily(raw, file, language) {
  const meta = readFrontmatter(raw);
  const body = stripFrontmatter(raw);
  const sentenceSection = between(body, "## Sentences", "## 오늘 바로 써볼 문장")
    || body.replace(/^# .+\n*/, "").split("## 오늘 바로 써볼 문장")[0].trim();
  const sentenceBlocks = sentenceSection.split(/\n(?=\d+\.\s)/).map((block) => block.trim()).filter(Boolean);
  const sentences = sentenceBlocks.map((block) => {
    const lines = block.split("\n").map((line) => line.trim()).filter(Boolean);
    const textLines = [];
    for (const line of lines) {
      if (line.startsWith("한국어:") || line.startsWith("고난이도") || line.startsWith("- ")) break;
      textLines.push(line.replace(/^\d+\.\s*/, ""));
    }
    const { reading, original } = splitReadingAndOriginal(textLines, language);
    const korean = (lines.find((line) => line.startsWith("한국어:")) || "").replace(/^한국어:\s*/, "");
    const expressions = extractBulletExpressions(block);
    return { reading, original, korean, expressions };
  });
  const practice = (between(body, "## 오늘 바로 써볼 문장", "").match(/-\s*(.+)/) || [])[1] || sentences[0]?.original || "";

  return {
    date: meta.date || dateFromFile(file),
    language,
    title: `${language} Daily Sentences - ${meta.date || dateFromFile(file)}`,
    practice,
    sentences,
    expressions: sentences.flatMap((sentence) => sentence.expressions),
    sourceFile: file
  };
}

function parseIssue(raw, file, language) {
  const meta = readFrontmatter(raw);
  const body = stripFrontmatter(raw);
  const issueSection = between(body, "## Issue", "## 내 의견 한 줄");
  const korean = (issueSection.match(/한국어:\s*(.+)/) || [])[1] || "";
  const expressions = extractBulletExpressions(issueSection);
  const sourceLines = issueSection.split("\n").filter((line) => line.trim().startsWith("- ["));
  const opinionSection = between(body, "## 내 의견 한 줄", "");
  const practice = (opinionSection.match(/-\s*(?:English|Japanese|Chinese):\s*(.+)/) || opinionSection.match(/-\s*한국어:\s*(.+)/) || [])[1] || korean;
  const lines = issueSection.split("\n").map((line) => line.trim()).filter(Boolean);
  const textLines = lines.filter((line) => !line.startsWith("한국어:") && !line.startsWith("고난이도") && !line.startsWith("-") && !line.startsWith("Sources:"));
  const { reading, original } = splitReadingAndOriginal(textLines, language);

  return {
    date: meta.date || dateFromFile(file),
    language,
    title: `${language} World Issue - ${meta.date || dateFromFile(file)}`,
    topic: meta.topic || "",
    practice,
    issueCards: [
      {
        title: meta.topic || "World Issue",
        reading,
        original: original || korean,
        korean,
        expressions
      }
    ],
    expressions,
    sources: sourceLines,
    sourceFile: file
  };
}

function readFrontmatter(raw) {
  const match = raw.match(/^---\n([\s\S]*?)\n---/);
  if (!match) return {};
  return Object.fromEntries(
    match[1].split("\n").map((line) => {
      const index = line.indexOf(":");
      if (index === -1) return null;
      return [line.slice(0, index).trim(), line.slice(index + 1).trim()];
    }).filter(Boolean)
  );
}

function stripFrontmatter(raw) {
  return raw.replace(/^---\n[\s\S]*?\n---\n*/, "");
}

function between(text, startHeading, endHeading) {
  const start = text.indexOf(startHeading);
  if (start === -1) return "";
  const bodyStart = start + startHeading.length;
  const end = endHeading ? text.indexOf(endHeading, bodyStart) : -1;
  return text.slice(bodyStart, end === -1 ? text.length : end).trim();
}

function extractBulletExpressions(block) {
  const expressions = [];
  let current = null;

  for (const rawLine of block.split("\n")) {
    const line = rawLine.trim();
    if (line.startsWith("- Hiragana:") || line.startsWith("- Pinyin:")) {
      if (current) expressions.push(normalizeExpression(current));
      const [label, ...rest] = line.replace(/^- /, "").split(":");
      current = { readingLabel: label.trim(), reading: rest.join(":").trim() };
      continue;
    }

    if (current && (line.startsWith("표현:") || line.startsWith("한국어:") || line.startsWith("활용:"))) {
      const [label, ...rest] = line.split(":");
      current[label.trim()] = rest.join(":").trim();
      continue;
    }

    if (!line.startsWith("- ")) continue;
    const value = line.replace(/^- /, "");
    if (value.startsWith("[")) continue;
    if (current) {
      expressions.push(normalizeExpression(current));
      current = null;
    }
    const [term, ...rest] = value.split(":");
    if (cleanLabel(term)) {
      expressions.push({ term: cleanLabel(term), meaning: cleanLabel(rest.join(":").trim()) });
    }
  }

  if (current) expressions.push(normalizeExpression(current));
  return expressions.filter((item) => item.term);
}

function normalizeExpression(item) {
  return {
    term: cleanLabel(item["표현"] || item.term || ""),
    reading: cleanLabel(item.reading || ""),
    meaning: cleanLabel(item["한국어"] || item.meaning || ""),
    note: cleanLabel(item["활용"] || "")
  };
}

function cleanLabel(value) {
  return String(value || "").replace(/^(Hiragana|Pinyin|표현|한국어|활용):\s*/, "").trim();
}

function splitReadingAndOriginal(textLines, language) {
  const lines = textLines.filter(Boolean);
  if ((language === "Japanese" || language === "Chinese") && lines.length >= 2) {
    return { reading: lines[0], original: lines[1] };
  }
  return { reading: "", original: lines[0] || "" };
}

function dateFromFile(file) {
  return (file.match(/\d{4}-\d{2}-\d{2}/) || [""])[0];
}
